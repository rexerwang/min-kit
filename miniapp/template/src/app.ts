import '@miniapp/components/dist/styles/index.css'
import './assets/styles/index.scss'
import './assets/polyfill'
import './icons'

import { useDebug } from '@miniapp/components'
import { go, logger } from '@miniapp/extends'
import { useError, useLaunch, usePageNotFound, useUnhandledRejection } from '@tarojs/taro'

import { Pages } from './app.route'
import authStore from './shared/store/auth'

function App({ children }: React.PropsWithChildren) {
  useDebug({
    user: {
      getToken: authStore.apis.getToken,
      async getUserInfo(reLogin?: boolean) {
        if (reLogin) await authStore.apis.login()
      },
    },
  })

  useLaunch((opts) => {
    logger.debug('#AppLaunch', opts)
  })

  usePageNotFound((res) => {
    logger.error('#PageNotFound', res)

    go.redirect(Pages.Home)
  })

  useUnhandledRejection(({ reason }) => {
    logger.error('#UnhandledRejection', reason)
  })

  useError((error) => {
    logger.error('#AppError', error)
  })

  return children
}

export default App
