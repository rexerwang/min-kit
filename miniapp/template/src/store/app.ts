import { combineStore, createPersistStore } from '@miniapp/store'

interface IState {
  name: string
  author: string
  version: string
  lastSignedIn: { at: number; scene: number } | undefined
}

const initialState: IState = {
  name: '@miniapp/template',
  author: 'rexerwang',
  version: '1.0.0',
  lastSignedIn: undefined,
}

// TODO: fix typings
const store = createPersistStore('appStore', initialState, (set: any) => ({
  lastSignIn(at: number, scene: number) {
    set((state) => {
      state.lastSignedIn = { at, scene }
    })
  },
}))

const appStore = combineStore(store, {
  apis: {
    lastSignIn: store.getState().lastSignIn,
  },
})

export default appStore
