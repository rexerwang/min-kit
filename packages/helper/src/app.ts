export * from './app/defineAppConfigChain'
export * from './app/defineRouteConfig'
