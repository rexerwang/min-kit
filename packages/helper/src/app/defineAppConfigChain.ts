import ConfigChain from './chain/ConfigChain'

type ConfigEnv = {
  mode: NodeJS.ProcessEnv['TARO_MODE']
  env: NodeJS.ProcessEnv['NODE_ENV']
}

/**
 * defineAppConfig by {@link ConfigChain}
 *
 * @requires `process.env.TARO_MODE`
 * @requires `process.env.NODE_ENV`
 */
export function defineAppConfigChain(cb: (chain: ConfigChain, env: ConfigEnv) => void): Taro.Config {
  const chain = new ConfigChain()
  cb(chain, { mode: process.env.TARO_MODE, env: process.env.NODE_ENV })
  return chain.get()
}
