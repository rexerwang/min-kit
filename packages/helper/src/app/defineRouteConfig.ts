function normalizeRoutePath<T extends object>(routeMap: T) {
  return Object.entries(routeMap).reduce((map, [key, route]) => {
    if (!/^[A-Za-z0-9]+$/.test(key))
      throw new Error(`Only letters or numbers are allowed as the subPackage key: ${key}`)

    if (typeof route === 'string') {
      map[key] = '/' + route
    } else {
      const pkg = key
        .split('')
        .map((c, i) => (i === 0 || /[a-z0-9]/.test(c) ? c : '-' + c))
        .join('')
        .toLowerCase()

      const root = `/${pkg}/`

      map[key] = Object.entries(route).reduce((pages, [k, path]) => {
        pages[k] = root + path
        return pages
      }, {})
    }

    return map
  }, {} as T)
}

type RouteRecord = Record<string, string | Record<string, string>>

type RouteConfig<T> = {
  /** the **relative** path routes */
  Routes: T
  /** the **absolute** path routes */
  Pages: T
}

/**
 * Define route paths. with the nested subPackages (optional)
 *
 * @param Routes the relative path routes.
 * If there are subPackages, use the subPackage root (PascalCase) as the key.
 *
 * @example
 * ```js
 * defineRouteConfig({
 *   // main package
 *   Home: 'pages/index/index',
 *
 *   // pkg-abc subPackage
 *   PkgAbc: {
 *     Home: 'pages/index/index',
 *   }
 * })
 * ```
 */
export function defineRouteConfig<T extends RouteRecord>(Routes: T): RouteConfig<T> {
  return { Routes, Pages: normalizeRoutePath(Routes) }
}
