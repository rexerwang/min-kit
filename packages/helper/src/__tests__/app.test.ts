import { defineAppConfigChain, defineRouteConfig } from '../app'

describe('app specs', () => {
  describe('defineAppConfigChain', () => {
    it('should set app.config when given `TARO_ENV` is weapp', () => {
      jest.replaceProperty(process.env, 'TARO_ENV', 'weapp')

      expect(
        defineAppConfigChain((chain, { env }) => {
          chain
            .entryPagePath('pages/index/index')
            .pages(['pages/index/index', 'pages/order/index'])
            .subPackage('pkg-abc')
            .pages(['pages/index/index'])
            .plugin('plugin-id', { provider: 'plugin-provider', version: '1.0.0' })
            .end()
            .wechat.debug(env === 'development')
            .plugin('plugin-id', { provider: 'plugin-provider', version: '1.0.0' })
            .darkMode()
            .permission('get location desc')
            .requiredBackgroundModes(['location'])
            .requiredPrivateInfos(['getLocation'])
            .skyline()
            .end()
            .window({
              navigationBarTitleText: 'min-kit',
              navigationStyle: 'custom',
            })
            .tabBar({
              color: '#000',
              selectedColor: '#000',
              backgroundColor: '#fff',
              list: [
                {
                  pagePath: 'pages/index/index',
                  iconPath: 'home.png',
                  selectedIconPath: 'home.active.png',
                  text: 'Home',
                },
                {
                  pagePath: 'pages/order/index',
                  iconPath: 'order.png',
                  selectedIconPath: 'order.active.png',
                  text: 'Order',
                },
              ],
            })
        }),
      ).toMatchSnapshot()
    })

    it('should throw error when not set pages', () => {
      expect(() =>
        defineAppConfigChain((chain) => {
          chain.window({ navigationBarTitleText: 'min-kit' })
        }),
      ).toThrow('Please set pages via `chain.pages()`')
    })
  })

  describe('defineRouteConfig', () => {
    it('should defineRouteConfig did work', () => {
      const { Pages, Routes } = defineRouteConfig({
        Home: 'pages/index/index',
        PkgAbc: {
          Home: 'pages/index/index',
        },
      })

      expect(Pages).toEqual({
        Home: '/pages/index/index',
        PkgAbc: {
          Home: '/pkg-abc/pages/index/index',
        },
      })

      expect(Routes).toEqual({
        Home: 'pages/index/index',
        PkgAbc: {
          Home: 'pages/index/index',
        },
      })
    })

    it('it should throw error when given invalid keys', () => {
      expect(() => defineRouteConfig({ a_b_c: 'pages/index/index' })).toThrow(/Only letters or numbers are allowed/)
    })
  })
})
