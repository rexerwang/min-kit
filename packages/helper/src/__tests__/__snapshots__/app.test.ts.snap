// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app specs defineAppConfigChain should set app.config when given \`TARO_ENV\` is weapp 1`] = `
{
  "componentFramework": "glass-easel",
  "darkmode": true,
  "debug": false,
  "entryPagePath": "pages/index/index",
  "lazyCodeLoading": "requiredComponents",
  "pages": [
    "pages/index/index",
    "pages/order/index",
  ],
  "permission": {
    "scope.userLocation": {
      "desc": "get location desc",
    },
  },
  "plugins": {
    "plugin-id": {
      "provider": "plugin-provider",
      "version": "1.0.0",
    },
  },
  "renderer": "skyline",
  "rendererOptions": {
    "skyline": {
      "defaultDisplayBlock": true,
    },
  },
  "requiredBackgroundModes": [
    "location",
  ],
  "requiredPrivateInfos": [
    "getLocation",
  ],
  "subPackages": [
    {
      "name": "pkg-abc",
      "pages": [
        "pages/index/index",
      ],
      "plugins": {
        "plugin-id": {
          "provider": "plugin-provider",
          "version": "1.0.0",
        },
      },
      "root": "pkg-abc",
    },
  ],
  "tabBar": {
    "backgroundColor": "#fff",
    "color": "#000",
    "list": [
      {
        "iconPath": "home.png",
        "pagePath": "pages/index/index",
        "selectedIconPath": "home.active.png",
        "text": "Home",
      },
      {
        "iconPath": "order.png",
        "pagePath": "pages/order/index",
        "selectedIconPath": "order.active.png",
        "text": "Order",
      },
    ],
    "selectedColor": "#000",
  },
  "window": {
    "navigationBarTitleText": "min-kit",
    "navigationStyle": "custom",
  },
}
`;
